### PR TITLE
feat: add rejectLate option

### PR DESCRIPTION
**Pull request by wraithgar**
Closes: https://github.com/isaacs/promise-call-limit/issues/1

### CODE REVIEW

**Comment by wraithgar** [comment]
I should also update the README, which I'll do if I get the indication this is gonna land.

File: index.js [hunk 1]
```diff
@@ -39,9 +43,17 @@ const callLimit = (queue, limit = defLimit) => new Promise((res, rej) => {
...
       results[c] = result
-      run()
+      if (!rejected)
+        run()
       return result
-    }, reject)
+    }, (er) => {
+      active --
+      reject(er)
+    }).finally(result => {
```
Comment (wraithgar) [inline index.js:52]:
```suggestion
    }).then(result => {
```

Realized this was still here from debugging. Should be a normal `then`

**Comment by isaacs** [comment]
This seems surprising. If you have `rejectLate` set, then shouldn't it process _all_ items, even if there's an early rejection? It looks like it's not actually doing that.

```js
const promiseCallLimit = require('./')

const pass = async () => new Promise(r => setTimeout(() => {
  console.log('pass')
  r()
}))
const fail = async () => new Promise((_, r) => setTimeout(() => {
  console.log('fail')
  r('nope')
}))

promiseCallLimit([
  pass,
  fail,
  pass,
  pass,
  pass,
  pass,
  pass,
], 2, true).then(console.log, console.error)

// Expect 6 passes logged, 1 fail
// Actual:
// pass
// fail
// pass
// nope
```

**Comment by isaacs** [comment]
Consider:

```patch
diff --git a/index.js b/index.js
index afba90d..ba9395a 100644
--- a/index.js
+++ b/index.js
@@ -36,22 +36,24 @@ const callLimit = (queue, limit = defLimit, rejectLate) => new Promise((res, rej
   const run = () => {
     const c = current++
     if (c >= queue.length) {
-      return resolve()
+      return rejected ? reject() : resolve()
     }
 
     active ++
     results[c] = queue[c]().then(result => {
       active --
       results[c] = result
-      if (!rejected)
+      if (!rejected || rejectLate)
         run()
       return result
     }, (er) => {
       active --
       reject(er)
-    }).finally(result => {
-      if (rejectLate && rejected && active === 0)
+    }).then(result => {
+      if (rejectLate && rejected && active === 0) {
+        console.error('rejectLate and rejected')
         return rej(rejection)
+      }
       return result
     })
   }

```

**Comment by isaacs** [comment]
> Yes, the idea is that if any in-flight promise rejects, no new ones are started, but all in-flight promises finish before that first rejection bubbles up. Any extra rejections are disregarded.

I see. That seems fine, and is what this does, but should definitely be documented, because it's easy to assume that rejectLate will behave like promise-all-reject-late, where _all_ promises are guaranteed to resolve before returning.

**Comment by isaacs** [comment]
Maybe instead of a `rejectLate` boolean, it should be an options object? I'm not sure how to best capture the different senses of "rejectLate" behavior, but they both seem reasonable to want, in different scenarios.

**Comment by wraithgar** [comment]
> Consider:

This makes the `t.notOk(failure2, 'slow rejection never ran')` test on line 88 fail.

The existing tests also never catch the entirety of `if (!rejected || rejectLate)`

**Comment by wraithgar** [comment]
I'd love to switch to an options object.  I opted for a non-breaking change for the initial PR.

**Comment by wraithgar** [comment]
Since we're in net new territory if folks want ALL the promises to run they really want a `Promise.allSettled` behavior and perhaps that would be the option name.

**Comment by wraithgar** [comment]
Took a stab at an opts object and updating the readme.

File: README.md [hunk 2]
```diff
@@ -10,18 +10,26 @@ const promiseCallLimit = require('promise-call-limit')
...
+  - rejectLate: if true, then any rejection will stop new promises from
+    being ran, and will wait for all currently running promises to
+    resolve (or reject) before finally rejecting with that first
+    rejection.  All subsequent rejections will be ignored.
 
-The reason for subtracting one is that presumably the main thread is taking
-up a CPU as well, so let's not be greedy.
+Whatever the limit is set to, only one less than that limit will be
+used.  Presumably the main thread is taking up a CPU as well, so let's
+not be greedy.
```
[the remark this answers is not in the document]
Comment (wraithgar) [reply README.md:32]: Yeah that's not right.  The "one less" is not being done at all, I typed that w/o looking at the for loop again.  We need to decrease the default by one.

**Comment by isaacs** [comment]
> ETA: Oh does this run all promises? That'd be a good implementation of "reject last" as I'm calling it.

Yeah, that's when I thought the intent was for `rejectLate` to run all the way to the end, we had a few rapid-fire comments pass midair there, I think 😅

**Comment by wraithgar** [comment]
It's entirely possible my initial assumption about how `rejectLate` would work is wrong.  I had assumed the behavior in this PR would be the sensible default but I can now see how `rejectLate` already has a pretty tight coupling to the `allSettled` behavior.

I also assumed this was the behavior we'd want in npm, but it's very likely I was wrong.

Ignoring my assumption so that we can try to get rid of the scope creep here, what do you think the minimum implementation of this should be here?

**Comment by wraithgar** [comment]
Rewrote the PR to discard the idea of stopping at the first rejection.

File: index.js [hunk 3]
```diff
@@ -5,20 +5,24 @@ const os = require('os')
 
 /* istanbul ignore next - version-specific workaround */
 const defLimit = 'availableParallelism' in os
-  ? os.availableParallelism()
-  : Math.max(1, os.cpus().length)
+  ? Math.max(1, os.availableParallelism() - 1)
+  : Math.max(1, os.cpus().length - 1)
 
-const callLimit = (queue, limit = defLimit) => new Promise((res, rej) => {
+const callLimit = (queue, { limit = defLimit, rejectLate } = {}) => new Promise((res, rej) => {
```
Comment (wraithgar) [inline index.js:11]: Breaking change

**Pull request closed without merging**